### PR TITLE
ESQL: Rework building evaluators

### PR DESCRIPTION
**Pull request by nik9000**
This reworks `EvaluatorMapper` to use a named interface instead of a `Function<Expression, EvalOperator.ExpressionEvaluator.Factory>`. It's shorter to type and it allows us to add extra information to the builder.

### CODE REVIEW

**Review by ivancea — APPROVED** [review]

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/mapper/EvaluatorMapper.java [hunk 1]
```diff
@@ -14,15 +14,17 @@
...
-
 import static org.elasticsearch.compute.data.BlockUtils.fromArrayRow;
 import static org.elasticsearch.compute.data.BlockUtils.toJavaObject;
 
 /**
  * Expressions that have a mapping to an {@link ExpressionEvaluator}.
  */
 public interface EvaluatorMapper {
+    interface ToEvaluator {
+        ExpressionEvaluator.Factory toEvaluator(Expression expression);
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/mapper/EvaluatorMapper.java:25]: bikeshed: I think the previous `apply` was easier on the eyes; `toEvaluator.toEvaluator(...)` looks a bit funny.

**Review by alex-spies — APPROVED** [review]
Thanks for zapping all the `Function<Expression, ExpressionEvaluator.Factory>`s, Nik!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/mapper/EvaluatorMapper.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/mapper/EvaluatorMapper.java:25]: Fair. I figured if we were adding another thing to the interface it'd be weird to still have the method called `apply`.

Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/mapper/EvaluatorMapper.java:25]: I think `apply` looks nicer too. What will the other method do? If the `apply` is the primary thing to call, that still sounds right to me.

Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/mapper/EvaluatorMapper.java]: :+1:

**Comment by nik9000** [comment]
backport robot hasn't picked this up in the past half hour. I'm doing it by hand.

**Pull request merged**